### PR TITLE
DHCPv6: add support for stateful DHCPv6 (RFC 8415)

### PR DESCRIPTION
**Pull request by francescolavra**
This adds supports for receiving IPv6 addresses from a DHCPv6 server. It implements sending of solicit, request, renew and rebind messages, and reception of advertise and reply messages.

### CODE REVIEW

File: src/core/ipv6/dhcp6.c [hunk 1]
```diff
@@ -272,7 +279,505 @@ dhcp6_stateless_enabled(struct dhcp6 *dhcp6)
...
+dhcp6_msg_finalize(u16_t options_out_len, struct pbuf *p_out)
+{
+  /* shrink the pbuf to the actual content length */
+  pbuf_realloc(p_out, (u16_t)(sizeof(struct dhcp6_msg) + options_out_len));
+}
+
+static void
+dhcp6_set_req_timeout(struct dhcp6 *dhcp6, u16_t initial_rt, u16_t max_rt)
+{
+  u32_t secs = (dhcp6->tries < 16) ? (initial_rt << (dhcp6->tries - 1)) : max_rt;
```
Comment (wjhun) [inline src/core/ipv6/dhcp6.c]: dhcp6->tries cannot be zero here, right? Maybe assert?

**Review by wjhun — APPROVED** [review]
I am only able to follow some of the protocol stuff, but otherwise it looks good to me and safe to merge. It would be good to send this upstream and get some more useful feedback from the lwIP maintainers.

File: src/core/ipv6/dhcp6.c (hunk 1 above)
Comment (francescolavra) [reply src/core/ipv6/dhcp6.c]: Correct, dhcp6->tries cannot be zero, I added an assert

**Comment by francescolavra** [comment]
Yes, I already sent the patchset upstream, at https://savannah.nongnu.org/patch/?10061

**Pull request merged**